src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm

```diff
@@ -1060,60 +1039,45 @@ - (void)controllerShuttingDown:(MTRDeviceController *)controller
     [_controllers removeObject:controller];
     os_unfair_lock_unlock(&_controllersLock);
 
-    // Snapshot the controller's fabric index, if any, before it clears it
-    // out in shutDownCppController.
-    __block FabricIndex controllerFabricIndex = controller.fabricIndex;
+    // Do the controller shutdown on the Matter work queue.
+    dispatch_sync(_chipWorkQueue, ^{
+        FabricIndex fabricIndex = controller.fabricIndex;
+        if (fabricIndex != kUndefinedFabricIndex) {
+            // Clear out out group keys for this fabric index, in case fabric
+            // indices get reused later.  If a new controller is started on the
+            // same fabric it will be handed the IPK at that point.
+            self->_groupDataProvider->RemoveGroupKeys(fabricIndex);
+        }
 
-    // This block runs either during sync dispatch to the Matter queue or after
-    // Matter queue shutdown, so it can touch any of our members without
-    // worrying about locking, since nothing else will race it.
-    auto sharedCleanupBlock = ^{
-        assertChipStackLockedByCurrentThread();
+        if (_otaProviderDelegateBridge) {
+            _otaProviderDelegateBridge->ControllerShuttingDown(controller);
+        }
 
         [controller shutDownCppController];
 
         self->_controllerBeingShutDown = nil;
         if (self->_controllerBeingStarted == controller) {
-            controllerFabricIndex = self->_nextAvailableFabricIndex;
             self->_controllerBeingStarted = nil;
         }
-    };
 
-    if ([_controllers count] == 0) {
-        dispatch_sync(_chipWorkQueue, ^{
+        // If there are no other controllers left, we can shut down some things.
+        if (_controllers.count == 0) {
             delete self->_operationalBrowser;
             self->_operationalBrowser = nullptr;
-        });
-        // That was our last controller.  Stop the event loop before it
-        // shuts down, because shutdown of the last controller will tear
-        // down most of the world.
-        DeviceLayer::PlatformMgrImpl().StopEventLoopTask();
-
-        if (_otaProviderDelegateBridge) {
-            _otaProviderDelegateBridge->Shutdown();
-            _otaProviderDelegateBridge.reset();
-        }
 
-        if (_otaProviderEndpoint != nil) {
-            [_otaProviderEndpoint unregisterMatterEndpoint];
-            [_otaProviderEndpoint invalidate];
-
-            [self removeServerEndpoint:_otaProviderEndpoint];
-
-            _otaProviderEndpoint = nil;
-        }
-
-        sharedCleanupBlock();
-    } else {
-        // Do the controller shutdown on the Matter work queue.
-        dispatch_sync(_chipWorkQueue, ^{
             if (_otaProviderDelegateBridge) {
-                _otaProviderDelegateBridge->ControllerShuttingDown(controller);
+                _otaProviderDelegateBridge->Shutdown();
```

Comment:
We used to do this before last-controller shutdown... But now we will do it after, no?  And that code uses the exchange manager and whatnot.  If the system state is shut down by then, won't those objects be dead after we do shutDownCppController?

Reply:
The `_otaProviderDelegateBridge->ControllerShuttingDown(controller);` is still there above, before we call shutDownCppController. I think this is a case of the diff making it look confusing.

Reply:
I realize the `ControllerShuttingDown()` bit is there.

But I am looking at the implementation of `Shutdown()` and it does:

```
        mExchangeMgr->UnregisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id);
```

Reply:
Ah got it, fixed.